### DISCUSSION TITLE
TELOPT IAC commands not sent through websocket

### DISCUSSION THREAD

**Discussion by torinwalker** [fluffos/fluffos#1096] [General]
Using MUDlet on telnet port shows full TELOPT / GMCP negotiation as would be expected. Using a well-established websocket client pointed at the websocket port, driver is not sending any IAC commands at all.

Are TELOPT IAC commands limited to telnet protocol only and not supported on websocket?

Using fluffos-20230904-3279863 and Dead Souls 3.9.

**Comment by michaelprograms** [comment]
This is the [current open ticket](https://github.com/fluffos/fluffos/issues/840) for adding this feature. Basically, the websocket client doesn't have support for telnet sequences per comments on [this issue](https://github.com/fluffos/fluffos/issues/1012).

**Reply by torinwalker** [reply]
Any suggestions on an approach? Don't mind adding it myself and kicking back the feature to you guys, but it's your code, so looking for best practices.

Have only had a cursory look, but the callback after handshake before passing control to the user object in ws-ascii.cc looks like a good place to put it, but data I've been sending (from the mudlib through receive) gets downgraded from binary to ascii (IAC 255 becomes 195, DO 253 becomes 193, etc. ) Some kind of filtering going on. Would need to experiment to see if it would even pass through the stack without getting chopped.

**Reply by michaelprograms** [reply]
Honestly I'm not too sure on the right approach, I'm not a pro at C++ or fluffos code architecture (yet) so can't really give any guidance, sorry.

I know fallentree mentioned he had done some preliminary work but I'm not sure if that's available anywhere. That said, if you do figure out something that works, I'm sure there's a bunch of people who would appreciate it.

**Reply by torinwalker** [reply]
I will reply to the last issue he dealt with and see if he has any ideas. Thank you Michael.

**Comment by torinwalker** [comment]
@thefallentree Per the request above, I am wondering if you can provide some general guidance on how you might implement TELOPT command support within the websocket protocol?

After connection is made in comm.cc::new_conn_handler(), it tests the port type and if telnet goes on to negotiate telnet options. If port is websocket, the handshake hasn't completed yet so it just returns. But later in the handshake callback in ws_ascii.cc, it doesn't continue by negotiating TELOPT options. As a result, TELOPTS and therefore GMCP, MXP, MSP, etc., etc. aren't supported by websockets.

I used ws_ascii_send to write a few simple negotiation commands in the ws_ascii handshake callback just before it calls on_user_logon, and it was successful - the client receives the commands and tries responding. Next I'll figure where to process incoming TELOPT commands (e.g. LWS_CALLBACK_RECEIVE) and process out TELOPTS, leaving whatever data to pass back to on_user_websocket_received.  My line of thinking is to duplicate telnet.cc as ws_telnet.cc and replace the 3rd-party send calls with ws_ascii_send calls. And then there's the fact that multiple connections all of which have a state, so it'll need a connection context. Plus commands may be split across calls/buffers, so state must be preserved (in the user?) as well. Not sure if/how to do this yet.

Anyway, before I go down this road too far, @michaelprograms  tells me you might have already started an implementation. If so, I'd like to hear your thoughts/design suggestions. Perhaps I could continue the work you started?

I have configured DS3.9 on top of fluffos_230904 and have an opportunity to use an amazing, fully featured client, but it uses TELOPTs and only talks websockets, so this is important enough to dedicate my time to fluffos, if you're open to contributions.

**Comment by thefallentree** [comment]
#1097 has largely solved this issue, give it a try! start a websocket connection with subprotocol specified as "telnet"!
